Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -1,113 +1,132 @@
-    public async Task<DafnyDocument?> UpdateDocumentAsync(DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
-      var documentId = documentChange.TextDocument;
-      while (_documents.TryGetValue(documentId.Uri, out var oldDocument)) {
-        cancellationToken.ThrowIfCancellationRequested();
-        if (documentId.Version < oldDocument.Version) {
-          _logger.LogDebug("skipping update of {DocumentUri} since the current version is newer (old={OldVersion} > new={NewVersion})",
-            documentId.Uri, oldDocument.Version, documentId.Version);
-          return oldDocument;
-        }
-        var mergedDocument = await _documentUpdater.ApplyChangesAsync(oldDocument, documentChange, cancellationToken);
-        if (_documents.TryUpdate(documentId.Uri, mergedDocument, oldDocument)) {
-          return mergedDocument;
-        }
+    public async Task<DafnyDocument> UpdateDocumentAsync(DidChangeTextDocumentParams documentChange) {
+      var documentUri = documentChange.TextDocument.Uri;
+      if (!documents.TryGetValue(documentUri, out var databaseEntry)) {
+        throw new ArgumentException($"the document {documentUri} was not loaded before");
       }
-      _logger.LogWarning("received update for untracked document {DocumentUri}", documentId.Uri);
-      return null;
+      if (documentChange.TextDocument.Version != databaseEntry.Version + 1) {
+        throw new InvalidOperationException($"the updates of document {documentUri} are out-of-order");
+      }
+      databaseEntry.CancelPendingUpdates();
+      var cancellationSource = new CancellationTokenSource();
+      var updatedEntry = new DocumentEntry(
+        documentChange.TextDocument.Version,
+        Task.Run(async () => await documentUpdater.ApplyChangesAsync(await databaseEntry.Document, documentChange, cancellationSource.Token)),
+        cancellationSource
+      );
+      documents[documentUri] = updatedEntry;
+      return await updatedEntry.Document;
     }
 
-    public async Task<DafnyDocument?> SaveDocumentAsync(TextDocumentIdentifier documentId, CancellationToken cancellationToken) {
-      if (VerifyOnSave) {
-        return await VerifyDocumentAsync(documentId, cancellationToken);
+    public async Task<DafnyDocument> SaveDocumentAsync(TextDocumentIdentifier documentId) {
+      if (!documents.TryGetValue(documentId.Uri, out var databaseEntry)) {
+        throw new ArgumentException($"the document {documentId.Uri} was not loaded before");
       }
-      if (_documents.TryGetValue(documentId.Uri, out var document)) {
+      var document = await databaseEntry.Document;
+      if (!VerifyOnSave) {
         return document;
       }
-      return null;
+      databaseEntry.CancelPendingUpdates();
+      var cancellationSource = new CancellationTokenSource();
+      var updatedEntry = new DocumentEntry(
+        document.Version,
+        RequiresOnSaveVerification(document)
```

Comment:
I have trouble following `SaveDocumentAsync`. I guess you're cancelling the document update and then restarting it because you have trouble only triggering a verification without also triggering parsing/resolution, due to how `TextDocumentLoader.loadRequests` works.

I would have expected something like:

```
if (VerifyOnSave) {
  var document = await _documents.GetDocumentAsync(request.TextDocument);
  if (document.SymbolTable.Resolved) {
    VerifyDocument(..) // <-- I guess this is the method you don't have.
  }
}
```

Is it difficult to still write such as `VerifyDocument` method?

Reply:
I refactored the `SaveDocumentAsync` method to better denote the single steps.
You are right, I am canceling any pending requests since I trigger a complete reload of the document. The reload behavior is the same as before, just that it couldn't cancel pending updates back then.
I've already been thinking about separating the verification from document loading to have more fine-granular over it. My goal would be to allow manually triggered verifications. But I think that the separation alone would already better fit in a dedicated PR.

Reply:
> I think that the separation alone would already better fit in a dedicated PR.

Makes sense. Could you add a comment explaining why it's currently the way it is?

Reply:
done